### PR TITLE
feat(backup): support crossplatform backup (CPB) POC [WPB-10575]

### PR DESCRIPTION
**Pull request by vitorhugods**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-10575" title="WPB-10575" target="_blank">StoryWPB-10575</a>  Cross Platform Backup: Write common backup / restore library
  </td></table>
  <br />
 

## Description

The goal is to unify the backup format between all clients, so a backup can be used in different devices/platforms.
It would also be a step in the direction of sharing chat history between clients. As this backup format can be shared from client A to client B.

## Checklist

- [X] mentions the JIRA issue in the PR name (Ex. [WPB-423])
- [X] PR has been self reviewed by the author;
- [X] Hard-to-understand areas of the code have been commented;
- [ ] If it is a core feature, unit tests have been added;

### This is a POC

Don't take everything seriously (especially the silly mistakes coming from a non TS/JS fluent person).

### Focus on src/script/backup/BackupRepository.ts

The directory `TEMP-crossplatform-backup` is only temporary and will _**not**_ live within wire-webapp repository. The idea is to publish a NPM artifact and remove these files from here. There's also some cleanup needed and most of these files will be removed from the artifact itself.

### Backup Logic

It's supposed to be exactly the same. The _only_ difference will be that instead of doing `JSON.stringify` and `TextEncode`, it will use this library to (de)serialize a common data format.

### What's actually expected from this PR?

Main feedback I'd like to have is about the general idea and the API itself. The interaction/contract between `MPBackupImporter` or `MPBackupExporter` and the Web client.
Does it feel idiomatic enough? How could I change the signature of functions to make it feel smoother?

### Silly example

| A conversation from Android. Exported from Android | This is the console logs after importing the backup file on Web |
| -------------------------------------| -------------------------------------|
|CPBackup1|CPBackup2|

Here is the silly backup file, FWIW. Rename it to `.desktop_wbu` in order to import on Web. I just made it a ZIP in order to upload on Github 😎

[WBX-vitor_wire-2024-11-04_16-19-18.zip](https://github.com/user-attachments/files/17620971/WBX-vitor_wire-2024-11-04_16-19-18.zip)

[WPB-423]: https://wearezeta.atlassian.net/browse/WPB-423?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by yamilmedina — APPROVED** [review]

**Review by przemvs — APPROVED** [review]

File: src/script/backup/BackupRepository.ts [hunk 1]
```diff
@@ -140,61 +117,36 @@ export class BackupRepository {
...
-
     const files: Record<string, Uint8Array> = {};
 
-    const stringifiedMetadata = JSON.stringify(metaData, null, 2);
-    const encodedMetadata = new TextEncoder().encode(stringifiedMetadata);
+    if (this.canceled) {
+      throw new CancelError();
+    }
+
+    // If the exported data is an Int8Array, it is a Multiplatform backup
```
Comment (tlebon) [inline src/script/backup/BackupRepository.ts]: can we align the naming within the codebase
```suggestion
    // If the exported data is an Int8Array, it is a cross-platform backup
```

File: src/script/backup/BackupRepository.ts [hunk 2]
```diff
@@ -129,8 +86,28 @@ export class BackupRepository {
...
     this.canceled = false;
 
+    const {
+      FEATURE: {ENABLE_CROSS_PLATFORM_BACKUP_EXPORT},
+    } = Config.getConfig();
+
     try {
-      const exportedData = await this._exportHistory(progressCallback);
+      let exportedData = null;
+      // If the feature flag is enabled, export the history as a Multiplatform backup
```
Comment (tlebon) [inline src/script/backup/BackupRepository.ts]:
```suggestion
      // If the feature flag is enabled, export the history as a cross-platform backup
```

File: src/script/backup/BackupRepository.ts [hunk 3]
```diff
@@ -308,21 +262,28 @@ export class BackupRepository {
...
-        const entities = JSON.parse(data);
-        return {
-          entities,
-          filename,
-        } as FileDescriptor;
+    // Check for Multiplatform backup
+    if (isMPBackup(files)) {
+      // Import Multiplatform backup
+      const mpbData = await importMPBHistoryToDatabase({
+        backupService: this.backupService,
```
Comment (tlebon) [inline src/script/backup/BackupRepository.ts]: MPB/ multiplatform (and now the imports have multiple names too, lets just use one naming for the futures sake)

File: src/script/backup/CrossPlatformBackup/AssetMetadata.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+// const isFileAsset = (contentType: string, infoObject: unknown): infoObject is FileAsset =>
+//   AssetContentType.isData(contentType) && hasNameProperty(infoObject);
+//const isVideoAsset = (contentType: string, infoObject: unknown): infoObject is VideoAsset =>
+//  AssetContentType.isVideo(contentType) && isObject(infoObject) && infoObject !== null && 'name' in infoObject;
+//const isAudioAsset = (contentType: string, infoObject: unknown): infoObject is AudioAsset =>
+//  AssetContentType.isAudio(contentType) && isObject(infoObject) && infoObject !== null && 'name' in infoObject;
+//const isTextAsset = (contentType: string, infoObject: unknown): infoObject is TexttAsset =>
+//  AssetContentType.isText(contentType) && isObject(infoObject) && infoObject !== null && 'name' in infoObject;
+//const isOtherAsset = (contentType: string, infoObject: unknown): infoObject is OtherAsset =>
+//  AssetContentType.isOther(contentType) && isObject(infoObject) && infoObject !== null && 'name' in infoObject;
```
Comment (tlebon) [inline src/script/backup/CrossPlatformBackup/AssetMetadata.ts:74]: im not sure id leave in potential future code, id rather leave a reference to a future ticket that contains the info.

File: src/script/backup/CrossPlatformBackup/CPB.helper.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import Dexie from 'dexie';
+
+import {ClientEvent} from 'src/script/event/Client';
+import {getLogger} from 'Util/Logger';
+
+import {MPBackup} from './CPB.library';
```
Comment (tlebon) [inline src/script/backup/CrossPlatformBackup/CPB.helper.ts]: MPBackup exported from CPB!? 🫠

**Pull request merged**